alpaca_trade_api/stream2.py

```diff
@@ -103,51 +86,113 @@ def __init__(self, key_id=None, secret_key=None, base_url=None):
+            self._ws = None
+
+    def _cast(self, channel, msg):
+        if channel == 'account_updates':
+            return Account(msg)
+        return Entity(msg)
+
+    async def _dispatch(self, channel, msg):
+        for pat, handler in self._handlers.items():
+            if pat.match(channel):
+                ent = self._cast(channel, msg['data'])
+                await handler(self, channel, ent)
+
+    def register(self, channel_pat, func, symbols=None):
+        if not asyncio.iscoroutinefunction(func):
+            raise ValueError('handler must be a coroutine function')
+        if isinstance(channel_pat, str):
+            channel_pat = re.compile(channel_pat)
+        self._handlers[channel_pat] = func
+        self._handler_symbols[func] = symbols
+
+    def deregister(self, channel_pat):
+        if isinstance(channel_pat, str):
+            channel_pat = re.compile(channel_pat)
+        self._handler_symbols.pop(self._handlers[channel_pat], None)
+        del self._handlers[channel_pat]
+
+
+class StreamConn(object):
+
+    def __init__(self, key_id=None, secret_key=None, base_url=None, data_url=None):
+        _key_id, _secret_key, _ = get_credentials(key_id, secret_key)
+        _base_url = base_url or get_base_url()
+        _data_url = data_url or get_data_url()
+
+        self.trading_ws = _StreamConn(_key_id, _secret_key, _base_url)
+        self.data_ws = _StreamConn(_key_id, _secret_key, _data_url)
+        self.polygon = polygon.StreamConn(_key_id +
+            '-staing' if 'staging' in _base_url else '')
+
+        self._handlers = {}
+        self._handler_symbols = {}
+
+        try:
+            self.loop = asyncio.get_event_loop()
+        except websockets.WebSocketException as wse:
+            logging.warn(wse)
+            self.loop = asyncio.new_event_loop()
+            asyncio.set_event_loop(self.loop)
+
+    async def _ensure_ws(self, conn):
+        if conn._handlers:
             return
+        conn._handlers = self._handlers.copy()
+        conn._handler_symbols = self._handler_symbols.copy()
+        await conn._connect()
 
-        ws_channels = []
-        polygon_channels = []
+    async def subscribe(self, channels):
+        '''Start subscribing to channels.
+        If the necessary connection isn't open yet, it opens now.
+        '''
+        trading_channels, data_channels, polygon_channels = [], [], []
         for c in channels:
             if c.startswith(('Q.', 'T.', 'A.', 'AM.',)):
                 polygon_channels.append(c)
+            elif c in ('trade_updates', 'account_updates'):
+                trading_channels.append(c)
             else:
-                ws_channels.append(c)
+                data_channels.append(c)
 
-        if len(ws_channels) > 0:
-            # Currently our streams don't support unsubscribe
-            # not as useful with our feeds
-            pass
+        if trading_channels:
+            await self._ensure_ws(self.trading_ws)
```

Comment:
I would prefer to make 3 channels async to each other. As it stands, this is sequential.

Reply:
Hmm but it is what it was, so maybe it's ok.